### PR TITLE
AIP-58: Add object storage backend for xcom

### PR DESCRIPTION
**Pull request by bolkedebruin**
This adds the possibility to store xcoms on a object storage supported backend.

This is an initial implementation and I am open for feedback (duh ;-) ). The BaseXCom is a bit messy and required some bug fixes. Hence, I have implemented the backend currently separated from BaseXCom. While functionally it is quite easy to merge both.

Notes:

* Made sure that custom serialization is now called when using `get_one`
* Added the `purge` function that is called when deleting / clearing xcoms so custom backend can purge xcom when required. This has also been reported as a bug: https://github.com/apache/airflow/issues/31774
* I have looked at FSMap that comes with fsspec, but decided that wasn't worth the effort and did not provide extra flexibility
* I could move this implementation to `common.io` but given that I'd like to merge it with the standard BaseXCom I though to put it where it is now. Can be discussed of course
* The state of BaseXCom leaves something to be desired

@potiuk @dstandish @hussein-awala @uranusjr 

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/config_templates/config.yml
Comment (potiuk) [inline airflow/config_templates/config.yml]: Nice!

File: airflow/config_templates/config.yml
Comment (potiuk) [inline airflow/config_templates/config.yml]: Would be nice to specify behaviour for `-1` - from default it looks like it is intended for `-1` to mean no threasholdm but also 0 could be one (0 feels a bit more intuitive, but if it's explicitly stated that `-1` makes all data stored in object storage, then -1 is a also good. It's of course logically correct (with exceeds and figuring out what happens when you have 0-sized object), but spelling it out explicitly would save the mental effort to figure it out by the user.

File: airflow/io/xcom.py
Comment (potiuk) [inline airflow/io/xcom.py]: I think we should have `missing_ok=True` here - this way  we handle purge also for low-threshold value here.

Comment (potiuk) [reply airflow/io/xcom.py]: Or just a case that somene else deleted it already (if the above is handled by the if from line 104)

File: airflow/io/xcom.py
Comment (potiuk) [inline airflow/io/xcom.py]: I believe you wanted to do `if not p.parent.exists()` ?  Technically speaking, p should NEVER exist here - excluding the collision case. 

And that leads to the question - should we handle collision here? 

I think it does not cost us much, so I'd say we should do smth like:

```python
while True:
   p = ObjectStoragePath(path) / f"{run_id}/{task_id}/{str(uuid.uuid4())}"
   if not p.exists():
       continue

if not p.parent.exists():
     ...
```

File: airflow/io/xcom.py
Comment (potiuk) [inline airflow/io/xcom.py]: There are two potential problems here:

1) there is quite an extra overhead for small values here - where retrieving of small value always involves extra storage call that could be avoided
2) there is a potential (small probablitly of course) - that there will be a storage object created externally and it might wreak havoc

Maybe another option would be to add a bool field in the Xcom table to distinquish local/remote? Mostly because of 1). I think it would be a way more performant approach for a number of cases. We've already had precedents with secret backends where reaching out to remote before checking things in the DB caused a lot of cost and network overhead. I think we should learn from that.

**Review by potiuk — COMMENTED** [review]
Made a first pass

File: airflow/config_templates/config.yml
Comment (bolkedebruin) [reply airflow/config_templates/config.yml]: I think from reading your response threshold might be a misnomer. -1 signifies that everything goes to the database. 0 signifies that everything goes to object storage as any size of a variable will be large than 0. So opposite what you wrote.

File: airflow/io/xcom.py
Comment (bolkedebruin) [reply airflow/io/xcom.py]: good point

File: airflow/io/xcom.py
Comment (bolkedebruin) [reply airflow/io/xcom.py]: good catch. I'll take a look

File: airflow/io/xcom.py
Comment (bolkedebruin) [reply airflow/io/xcom.py]: I don't think the overhead is really there. Small values will not result in that extra call, because the key won't be valid and it will raise a ValueError and object storage won't be touched. 

On your no 2 we store a reference to the path now in xcom if it has been written out to object storage. I think that works the same as you are suggesting and a schema change make it better.

Basically what we do now is:

Serialize value 
 if above > threshold then store in objectstorage and store path reference in db
 else store value in db

The reference tells us whether it is in the db or on object storage. We do need to reach out to check.

Comment (potiuk) [reply airflow/io/xcom.py]: Yeah. I think you are right, but It's a bit relying on implicit behaviour in this case. I think it should be at least described in a comment here that we know about it and that was designed and not accidental feature

File: airflow/config_templates/config.yml
Comment (potiuk) [reply airflow/config_templates/config.yml]: Then it's even more sign that it deserves spelling out expliciltly what it does.

Comment (bolkedebruin) [reply airflow/config_templates/config.yml]: Yep. Done :-)

File: airflow/io/xcom.py
Comment (bolkedebruin) [reply airflow/io/xcom.py]: Okido will add that.

File: docs/apache-airflow/core-concepts/xcoms.rst
Comment (jscheffl) [inline docs/apache-airflow/core-concepts/xcoms.rst]: Can you add/use an example that is available in Airflow production docker per default? Else many users will directly get a bleedy nose if they want to test "cool features" broken by default.

File: airflow/models/xcom.py
Comment (hussein-awala) [inline airflow/models/xcom.py:658]: 👍

File: airflow/io/xcom.py
Comment (jscheffl) [inline airflow/io/xcom.py]: Two things I am not sure:

1. Do we need to pass the type of compression here - or is this auto-detected while reading?
2. If compression is passed, what happens if the user changed the compression configuration after some time. Is the compression type be kept and auto-detected?

**Review by hussein-awala — COMMENTED** [review]
It looks good! I'm just wondering if we should move the introduced XCom backend and its configurations to the `common.io` provider 🤔

File: airflow/io/xcom.py
Comment (jscheffl) [inline airflow/io/xcom.py]: I am not sure but would it make sense to have DAG id also in the path? Then if a DAG is deleted, housekeeping might be easier? But you can convinec me of anything good as well.
```suggestion
                p = ObjectStoragePath(path) / f"{dag_id}/{run_id}/{task_id}/{str(uuid.uuid4())}"
```

**Review by jscheffl — COMMENTED** [review]
Some comments. Looks good overall. Not tested by running though, was just reading.

File: airflow/io/xcom.py
Comment (bolkedebruin) [reply airflow/io/xcom.py]: That might be okay, I though dag_id could be None, but it seems it cannot (the tests on BaseXCom do use None)

File: airflow/io/xcom.py
Comment (bolkedebruin) [reply airflow/io/xcom.py]: I can set it to `infer` here.

File: docs/apache-airflow/core-concepts/xcoms.rst
Comment (bolkedebruin) [reply docs/apache-airflow/core-concepts/xcoms.rst]: Sure. I'll change it to zip.

**Comment by bolkedebruin** [comment]
> It looks good! I'm just wondering if we should move the introduced XCom backend and its configurations to the `common.io` provider 🤔

@hussein-awala as mentioned in the topic start we can two multiple ways:

1. Merge this in into BaseXCom
2. Keep it separate

If we chose 2 then it makes sense to move it into common.io (with all the extras that it requires cause it is a 'new' capability for a provider). 

As it is backwards compatible and by default does the same as BaseXCom (storing data in the db) option 1 might make more sense.

File: airflow/io/xcom.py
Comment (bolkedebruin) [reply airflow/io/xcom.py]: Done

File: docs/apache-airflow/core-concepts/xcoms.rst
Comment (bolkedebruin) [reply docs/apache-airflow/core-concepts/xcoms.rst]: gzip is now used

**Comment by potiuk** [comment]
> If we chose 2 then it makes sense to move it into common.io (with all the extras that it requires cause it is a 'new' capability for a provider).

I think I''d support @hussein-awala 's idea here. There is a great feature which is non-obvious when thing are added in provider, namely dependency independence and ability to upgrade/downgrde separately from Airflow. And it's not much different than what we already have with the other secret backends - which are also implemented in providers, so this one would just follow the suite, and turns `common.io` into actualy pretty useful provider.

Binding such "auxiliary" functionality into Airflow core is also against our "Airflow-as-a-platform" approach. Without my suggestions applied (where I considered adding a DB change to make it clearer what is local and what remote), in my view this one simply builds on top of the Public API that we already have and there is no particular reason it should be in "core" airflow.

**Comment by bolkedebruin** [comment]
> > If we chose 2 then it makes sense to move it into common.io (with all the extras that it requires cause it is a 'new' capability for a provider).
> 
> I think I''d support @hussein-awala 's idea here. There is a great feature which is non-obvious when thing are added in provider, namely dependency independence and ability to upgrade/downgrde separately from Airflow. And it's not much different than what we already have with the other secret backends - which are also implemented in providers, so this one would just follow the suite, and turns `common.io` into actualy pretty useful provider.
> 
> Binding such "auxiliary" functionality into Airflow core is also against our "Airflow-as-a-platform" approach. Without my suggestions applied (where I considered adding a DB change to make it clearer what is local and what remote), in my view this one simply builds on top of the Public API that we already have and there is no particular reason it should be in "core" airflow.

Sure. I can relate to that. What about fully integrating this into BaseXCom? So not making it an auxiliary as I stated wasn't my intention. While working with BaseXCom I noticed that it would be good to refactor it a clean it up. By expanding its use into a provider we make the API more set in stone (I'm aware some of it is). 

Note that your suggestion of adding the extra field could create an issue for people that are currently using a custom xcom backend. They might already store it non-locally and the pattern is to store a reference in the value field. How to provide a proper migration path is then the question.

**Comment by potiuk** [comment]
Yeah. My suggestion was bad idea and you pointed out it's not needed indeed. And yes I thought that the change could actually be split In two : Base XCOM change for small/big values for all custom backends + common.io provider implementation if that's what you mean :) . Still the implementaiton of the Io custom backend should be - the more I think about it - in the provider even in this case.

**Comment by hussein-awala** [comment]
> > > If we chose 2 then it makes sense to move it into common.io (with all the extras that it requires cause it is a 'new' capability for a provider).
> > 
> > 
> > I think I''d support @hussein-awala 's idea here. There is a great feature which is non-obvious when thing are added in provider, namely dependency independence and ability to upgrade/downgrde separately from Airflow. And it's not much different than what we already have with the other secret backends - which are also implemented in providers, so this one would just follow the suite, and turns `common.io` into actualy pretty useful provider.
> > Binding such "auxiliary" functionality into Airflow core is also against our "Airflow-as-a-platform" approach. Without my suggestions applied (where I considered adding a DB change to make it clearer what is local and what remote), in my view this one simply builds on top of the Public API that we already have and there is no particular reason it should be in "core" airflow.
> 
> Sure. I can relate to that. What about fully integrating this into BaseXCom? So not making it an auxiliary as I stated wasn't my intention. While working with BaseXCom I noticed that it would be good to refactor it and give it a clean-up. By expanding its use into a provider we make the API more set in stone (I'm aware some of it is).
> 
> Why do I think it is better to have it in core? I think the use of custom xcom backends is limited. The bugs in de-serialization and non purging of values from custom backends should have let to more bug reports imho. Given that most patterns consider using a custom backend to store things on object storage we take that need away by implementing this. And yes, I understand that you can store things on redis or alike as well. What I mean with this is that it is a point to consider.
> 
> Note that your suggestion of adding the extra field could create an issue for people that are currently using a custom xcom backend. They might already store it non-locally and the pattern is to store a reference in the value field. How to provide a proper migration path is then the question.

I reviewed it again, I think you can just move the class `XComObjectStoreBackend` to the provider and keep the bug fix in the base XCom backend (you don't need to separate that in two PRs if I'm not mistaken), but you will need to add a check for the Airflow version in the module (>=2.8.2) similar to what we do for min Airflow version:
https://github.com/apache/airflow/blob/b15d5578dac73c4c6a3ca94d90ab0dc9e9e74c9c/airflow/providers/common/sql/__init__.py#L32-L42

**Comment by bolkedebruin** [comment]
@hussein-awala sure I can do that. Is that also what we want? Shouldn't we bump the requirements of the provider? I understand that this forces people to upgrade but the functionality won't be available or just partially? And does it make sense to have it more fine grained than the provider level? Seems to defeat the purpose of the provider minimum requirements a bit.

**Comment by hussein-awala** [comment]
> @hussein-awala sure I can do that. Is that also what we want? Shouldn't we bump the requirements of the provider? I understand that this forces people to upgrade but the functionality won't be available or just partially? And does it make sense to have it more fine grained than the provider level? Seems to defeat the purpose of the provider minimum requirements a bit.

The check is not for the whole provider, just for the module (or the class of the new XCom backend). If we keep the feature in the core package, the users will not be able to use it before 2.9.0, but after moving it to the provider and consider the core change as a bug fix, they will be able to use it starting from 2.8.2.
If you believe that the XCom backend will work without the core changes, and they will just have a problem with the delete, then maybe we can skip adding the check.

File: airflow/config_templates/config.yml
Comment (ephraimbuddy) [inline airflow/config_templates/config.yml]:
```suggestion
          database. 0 means always store in object storage. Any positive number means
```

File: airflow/config_templates/config.yml
Comment (uranusjr) [inline airflow/config_templates/config.yml]: I wonder if this should be more general. Having two configs that only make sense if `xcom_backend` is set to a specific backend seems awkward. I wonder if we can make the backend take arguments directly… One quick idea is to provide a backend _factory_ in Airflow instead, and tell the user to actually create the backend class dynamically…

```python
# $AIRFLOW_HOME/plugins/custom_xcom.py

CustomBackend = create_objectstorage_backend(
    storage_path="s3://conn_id@bucket/path",
    threshold=1_000_000,
    compression="gz",
)
```

```ini
# airflow.cfg
xcom_backend = custom_xcom.CustomBackend
```

Comment (bolkedebruin) [reply airflow/config_templates/config.yml]: Sure, of course I had in mind to make it part of core and the default. Is there any that does the factory pattern? I don't think it should reside in plugins as it should be easy to reach / config.

Comment (jscheffl) [reply airflow/config_templates/config.yml]: I like it the way it is at the moment - just enabling it by configuration is great. No need to inject custom code for a factory needed or a plugin.

**Review by jscheffl — APPROVED** [review]
I like it as it is - but before merge, other opinions needed!

File: airflow/config_templates/config.yml
Comment (potiuk) [inline airflow/config_templates/config.yml]: This should be moved to provider.yaml, since the storage backend is now there.

File: docs/apache-airflow/core-concepts/xcoms.rst
Comment (potiuk) [inline docs/apache-airflow/core-concepts/xcoms.rst:59]: This should be HowTo in provider docs as well.

File: airflow/providers/common/io/xcom/__init__.py
Comment (potiuk) [inline airflow/providers/common/io/xcom/__init__.py]: No need for this and __version__ below - those are only needed at the "package" level (and this file is generatel autoamatically there)

Comment (potiuk) [reply airflow/providers/common/io/xcom/__init__.py]: At most it should import version from pakage abve it is needed in the message below (but I think it is not needed).

File: airflow/providers/common/io/xcom/__init__.py
Comment (potiuk) [inline airflow/providers/common/io/xcom/__init__.py]: Here we do not require airflow 2.9.0 for the whole provider, just for XCom Feature of it . so the mesage should be

"The Xcom backend featuer in common.io provider needs ...."

Comment (potiuk) [reply airflow/providers/common/io/xcom/__init__.py]: Seems that the only reason why we need >= 2.9.0 is the new `purge` method.

Suggestion. Maybe we can really avoid that (and remove this file altogether. 

It looks like the only compatibility problem is lack of `purge` implementation in delete.  But for Airflow <2.9.0 we can easily override the whole delete method  (until we bump min airflow version to 2.9.0).

How about we (until min airflow version 2.9.0 for the whole provider) override delete method with a copy of the one from `arflow` - that would make it possible to use common.io.xcom.backend even in Airflow 2.6 

We could easily just implement delete method in `XComObjectStoreBackend` for example:

```python
    # TODO: Remove this method when we bump min_airflow_version to 2.9.0
    @classmethod
    @provide_session
    def delete(cls, xcoms: XCom | Iterable[XCom], session: Session) -> None:
        """Delete one or multiple XCom entries."""
       if ( if packaging.version.parse(packaging.version.parse(airflow_version).base_version) >=packaging.version.parse(
    "2.9.0") 
            return super.delete(xcoms=xcoms, session=session)
       if isinstance(xcoms, XCom):
            xcoms = [xcoms]
        for xcom in xcoms:
            if not isinstance(xcom, XCom):
                raise TypeError(f"Expected XCom; received {xcom.__class__.__name__}")
            XComObjectStoreBackend.purge(xcom, session)
            session.delete(xcom)
        session.commit()
```

I **think** that would work and that would show the REAL power and reason why want to move the XCom backend to provider - because it would give the XCom Backend to Airflow 2.8 users.

Comment (potiuk) [reply airflow/providers/common/io/xcom/__init__.py]: Ah no, also clear method would have to be copied. Probably not worh it.

Comment (potiuk) [reply airflow/providers/common/io/xcom/__init__.py]: Likely to prone to errors/easy to break.

Comment (hussein-awala) [reply airflow/providers/common/io/xcom/__init__.py]: > Here we do not require airflow 2.9.0 for the whole provider, just for XCom Feature of it . so the mesage should be

it's just applicable to `common.io.xcom` package, not the whole provider.

However, @ephraimbuddy the core change is a bugfix, will it be possible to add it to 2.8.2? if so, we need to update the minimum version in this module.

File: docs/apache-airflow/core-concepts/xcoms.rst
Comment (bolkedebruin) [reply docs/apache-airflow/core-concepts/xcoms.rst:59]: Will do

File: airflow/config_templates/config.yml
Comment (bolkedebruin) [reply airflow/config_templates/config.yml]: Yeah, I think the same. I'd like to keep it. There is no real harm and common.io is 'close' enough to core.

Comment (bolkedebruin) [reply airflow/config_templates/config.yml]: Ah okay, didn't know that was possible.

File: airflow/providers/common/io/xcom/__init__.py
Comment (bolkedebruin) [reply airflow/providers/common/io/xcom/__init__.py]: Haha I tried te be careful on the semantic versioning ;-). Indeed 'purge' is used at multiple places. One of the reasons I'd like to cleanup BaseXCom. We can go both ways here, 2.8.2 and treat it as a bug fix 'missing API component' or a change of API and make it 2.9.0.

Let me know: @ephraimbuddy what you prefer.

Comment (potiuk) [reply airflow/providers/common/io/xcom/__init__.py]: I like @hussein-awala 's way of thinking here :)

File: airflow/config_templates/config.yml
Comment (potiuk) [reply airflow/config_templates/config.yml]: That was my `vacation` pet-project for 2.7 release :)

File: airflow/providers/common/io/xcom/__init__.py
Comment (ephraimbuddy) [reply airflow/providers/common/io/xcom/__init__.py]: Would it be possible to make it two PRs, separating the core changes from the provider change? That way, it would be easier for users to know that it's a bug fix by merely reading the title.

Comment (bolkedebruin) [reply airflow/providers/common/io/xcom/__init__.py]: While possible, I'm not sure if that makes sense (and I am not looking forward to the work involved tbh). No one can make use of the 'bug fix'. I.e. it is a missing feature of the API and it will require a code change on the side of custom implementations to make use of it. It is a bug fix in the sense that you would expect the API to have it as it requires manual clean up on the side of the custom implementation now. 

In other words I'd like to keep it combined and just decide if the whole thing targets 2.9.0 or 2.8.2.

Comment (ephraimbuddy) [reply airflow/providers/common/io/xcom/__init__.py]: Ok. I'm okay with whichever one. I will remember it if marked 2.8.2

Comment (potiuk) [reply airflow/providers/common/io/xcom/__init__.py]: Yeah. Better to keep those in one PR. I think - contrary to  natural thinking - this is precisely the thing that is better when you have monorepo - that you **can** introduce changes to  multiple independent parts at the same time, and while resulting setup is more complex (because you have to manage back-compatibility and dependencies as we do here) generally development of it is quite a bit easier, because you effectively work on a combined change even if your change comes to two different components within the monorepo.

And it's not a problem at all in our release process either - there is completely separate release process and marking "PR" as being part either of the provider or or airflow release - the same PR will appear in both at some point of time and we can literally decide at the last moment before relasing provider whethere we mark the change also cherry-pickable to 2.8 branch.

This is the TRUE power of monorepo - development of such combined features is WAY easier.

Comment (potiuk) [reply airflow/providers/common/io/xcom/__init__.py]: So basically - as we use monorepo, there is virtually no benefit of splitting such PRs (and no need either).

Comment (potiuk) [reply airflow/providers/common/io/xcom/__init__.py]: Shall we change it to 2.8.2 then @bolkedebruin ?

File: airflow/config_templates/config.yml
Comment (uranusjr) [reply airflow/config_templates/config.yml]: The closest precedent we have is probably executors? We can create a new config section `[objectstorage_xcom]` that allows setting configs specific to the XCom backend. Although the section would be injected by a provider in the executor’s case, while we don’t have an equivalent here.

Comment (potiuk) [reply airflow/config_templates/config.yml]: But those config are already moved to the provider config :) .

File: airflow/providers/common/io/xcom/__init__.py
Comment (bolkedebruin) [reply airflow/providers/common/io/xcom/__init__.py]: Ill target 2.8.2

File: airflow/providers/common/io/xcom/backend.py
Comment (dirrao) [inline airflow/providers/common/io/xcom/backend.py:137]: I believe we are missing calling the BaseXCom serializer here.

**Review by dirrao — CHANGES_REQUESTED** [review]
Minor comment.

File: airflow/providers/common/io/xcom/backend.py
Comment (bolkedebruin) [reply airflow/providers/common/io/xcom/backend.py:137]: Mmm no not really, as s_val will be equal to what BaseXCom.serializer spits out but probably should call it here and move s_val to the if block.

update: which we cannot do as we need the len of the serialized value
update: fixed.

update: reverted this as we cannot use the same deserialize method from BaseXcom. Explained in file.

**Comment by jedcunningham** [comment]
(@bolkedebruin, moving this back to 2.9 as we won't ship this in a core patch release)

**Comment by bolkedebruin** [comment]
@jedcunningham it was agreed to have this in 2.8.2 as we consider this a bug fix more than a feature release. What has changed?

**Comment by jedcunningham** [comment]
Sorry, that decision was in a collapsed discussion I didn't expand :) It doesn't "feel" like a bugfix to me, but I'm happy to step back on this. @ephraimbuddy is the rm ultimately.

**Comment by bolkedebruin** [comment]
O yes the "expansion bug" ;-). I'm fine either way.  Targeting 2.8.2 will require a code change.

**Comment by cmarteepants** [comment]
I want to promote this PR - just confirming that it's going out in 2.8.2 for real? I agree with @jedcunningham that this feels like a new feature to me, but won't object to people getting it early.

**Comment by potiuk** [comment]
> @ephraimbuddy @bolkedebruin - I want to promote this PR - just confirming that it's going out in 2.8.2 for real? I agree with @jedcunningham that this feels like a new feature to me, but can never object to people getting something early.

Side comment: We can sneakily pretend it's a bug-fix (as if - it should have behaved  this way already, we just have not realized that. 


We had a lot of discussion on semver and classifying things and I am always in a position that SemVer is about the "intent" not technically breaking things, or technically adding new methods and fields and we can make individual decisions on some features or changes to classify them differently (mostly based on our assesment on how like it is things will get broken or how much this new feature is really "new") 

I think people take "compatibility" far too seriously and far too "technically" :).

**Comment by potiuk** [comment]
> Targeting 2.8.2 will require a code change.

Just chenging >= 2.9.0 right ? That's the only thing to change as far as I understand @bolkedebruin ?

**Comment by bolkedebruin** [comment]
@potiuk the code assumes >= 2.9.0 now. So no code change is required for targeting 2.9.0 from my point of view.

**Comment by bolkedebruin** [comment]
Fine to me!

**Pull request merged**
